### ISSUE TITLE
Named specification subsets

### ISSUE THREAD

**Issue by bitagoras** [thebuzzmedia/universal-binary-json#65]
Thanks to all of you for all the great ideas on a simple binary format. I always missed something like this. But I am a little bit afraid that UBJSON will stay too simple and restricted for me. I have read many of your contributions in this forum and I understand all arguments for simplicity. Simplicity is what made JSON so successfull. But I need some more features for very big numerical files (gigabyte data). So I tried to start my own project for a more versatile format, just for feassibility. It should keep simplicity for simple data but offer more features if needed. I would call it something like "Universal Binary Notation (UBN)", since it should not be restricted to JSON and java script. I would like to include a couple of additional features. For example:
* fast random access in very large files 
* small array type headers for short arrays
* complex structured data with minimal space
* nd-array support
* map/dict-support (objects with keys other data than string)

The last point is important for people working with python. Python includes json as a native datatype but on the other hand python data types can be more general (maps / dict objects with integers or even tuples = sparse matrix). These are examples of important features from outside the java world.

Now, I would not mention all this here. But I think that this new format would have only one disadvantage compared to UBJSON - it wouldl loose its simplicity. On the other hand a JSON representation can be a subset of it while keeping its simplicity. So my idea is a specification with *named subsets*, including one name for JSON. So the format is complex but some subsets are still simple. This would allow to create one single library that supports more and more subsets as milestones. At the end one single viewer for the full format specification could be created to view all kinds of format subsets, even very ridiculous simple subsets that are even simpler than json. All this is done by just giving subsets a specific name. This would unify the contradictory demands of simple and feature-rich specifications.

Maybe you don't like this idea of a more complex and versatile format. And it is maybe too late to adapt UBJSON to become a subset of something bigger like this. In this case please just have a look in my specifications if you find something usefull for UBJSON. It shows another example for the header definition for repeated data. I also try to introduce very short header types for small arrays and short strings to keep the header-to-data ratio small. A strategy to keep this large number of letter definition clean is the alphabetical order and a certain match between upper and lower case letters. 

Here is my current state of this idea (fast random access features are not included yet):
    
    Universal Binary Notation (UBN)
    -----------------------------------
    
    Specification Feature Subsets:
    ------------------------------
    A: Arrays container for values
    O: Objects container for key/value pairs
    S: Strings
    Q: Quick strings
    E: File with one element and no container
    R: Repeated elements with common type
    U: Short vectors up to 255 elements
    V: Vectors
    M: Maps with keys other than string
    N: Numeric types 
    H: High precision numeric types
    B: Boolean Type
    I: Size and offset information for fast random access (TBD)
    L: Links (TBD)
    
    Format Specification Subsets:
    -----------------------------
    Simple:  AON7N8S1        Very simple subset for arrays and objects with only 3 types: integer32, float64 and strings up to 255 byte
    Log:     AOSQERUV-N-B--  For logging applications (e.g. 2d tables with column names in the first row and fixed structured data types)
    JSON:    AOSQERUV-NHB-- Specification for a unique binary representation of JSON structures
    Full:    AOSQERUVMNHBIL  Full specification with maps and features for fast random access

    No   Symbol      Rule                                Description
    ---------------------------------------------------------------------------------------------------------------------------
    A0   FILE      = ARRAY                               Content of the file
    O0               OBJECT
    E0               ELEMENT                             File with single element
    A1   ARRAY     = [[] VALUES []]                      Array container
    O1   OBJECT    = [{] PAIRS [}]                       Container for key/value pairs
    O2   PAIRS     = PAIR
    O3               PAIRS PAIR
    A2   VALUES    = ITEMS
    A3               VALUES ITEMS
    A4   ITEMS     = VALUE                               Items inside an array
    R0               REPEAT(VALUE)
    R1   REPEAT(X) = [<] TYPE(X) [>] UINT DATA(X)        Type definition for repeated data of the same type
    O4   PAIR      = KEY VALUE
    R2               REPEAT(PAIRS)
    O5   KEY       = STRING                              Key (member name or index)
    M0               UINT
    M1               SCALAR
    M2               SHORTVEC
    M3               BOOL
    A5   VALUE     = ARRAY                               Value or complex object
    O6               OBJECT
    A5               ELEMENT
    R3               [!] VALUE                           Starts flagging all bytes in VALUE as payload data, to be moved to DATA(X)
    R4               [?] VALUE                           Starts flagging all bytes in VALUE as non-payload, to be moved to TYPE(X)
    N0   ELEMENT   = UINT
    A6               SCALAR
    Z0               ZERO
    N1               BOOL
    S0               STRING
    U0               SHORTVEC
    V1               VECTOR
    N2   UINT      = [u] UINT8                           Unsigned integer 8 bit
    N3               [v] UINT16                          Unsigned integer 16 bit
    N4               [w] UINT32                          Unsigned integer 32 bit
    H0               [x] UINT64                          Unsigned integer 64 bit
    N5   SCALAR    = [i] INT8                            Signed integer 8 bit
    N6               [j] INT16                           Signed integer 16 bit
    N7               [k] INT32                           Signed integer 32 bit
    H1               [l] INT64                           Signed integer 64 bit
    N8               [c] FLOAT32                         Single precision float
    N9               [d] FLOAT64                         Double precision float
    H2               [e] FLOAT128                        Quadruple precision float
    H3               [h] UINT8 n*BYTE                    Short high precision decimal float string
    H4               [H] UINT n*BYTE                     Long high precision decimal float string
    U0   SHORTVEC  = [I] UINT8 n*INT8                    Short array of signed integer 8 bit
    U1               [J] UINT8 n*INT16                   Short array of signed integer 16 bit
    U2               [K] UINT8 n*INT32                   Short array of signed integer 32 bit
    U3               [L] UINT8 n*INT64                   Short array of signed integer 64 bit
    U4               [U] UINT8 n*UNIT8                   Short array of unsigned integer 8 bit
    U5               [V] UINT8 n*UINT16                  Short array of unsigned integer 16 bit
    U6               [W] UINT8 n*UINT32                  Short array of unsigned integer 32 bit
    U7               [C] UINT8 n*FLOAT32                 Short array of single precision float
    U8               [D] UINT8 n*FLOAT64                 Short array of double precision float
    U9               [E] UINT8 n*FLOAT128                Short array of quadruple precision float
    B0   BOOL      = [b] BOOLEAN                         Bool (0x00 = false, 0xFF = true)
    B1               [t]                                 True
    B2               [f]                                 False
    Z1   ZERO      = [z]                                 Zero / null / none
    Q0   STRING    = [0]                                 String of length 0
    Q1               [1] 1*BYTE                          String of length 1
    Q2               [2] 2*BYTE                          String of length 2
    Q3               [3] 3*BYTE                          String of length 3
    Q4               [4] 4*BYTE                          String of length 4
    Q5               [5] 5*BYTE                          String of length 5
    Q6               [6] 6*BYTE                          String of length 6
    Q7               [7] 7*BYTE                          String of length 7
    Q8               [8] 8*BYTE                          String of length 8
    Q9               [9] 9*BYTE                          String of length 9
    S1               [s] UINT8 n*BYTE                    Short string
    S2               [S] UINT n*BYTE                     Long string
    V2   VECTOR    = [a] UINT8 TYPE(E) n*DATA(E)         Short array of any type with E = ELEMENT
    V3               [A] UINT TYPE(E) n*DATA(E)          Long array of any type with E = ELEMENT
    
    
    Notes:
        Brackets [ ] stands for one-byte ascii charactes, which represent terminal symbols.
        TYPE(X) := Everything in x without any payload data (except for values beginning with [!])
        DATA(X) := Only the payload data in X (except for values beginning with [?])
    
    Examples:
    
    1) String:
        "hello"
        = [5] [h] [e] [l] [l] [o]
    
    2) 3d vector of type uint8:
        [10, 200, 255]
        = [U] [uint8:3] [uint8:10] [uint8:200] [uint8:255]
    
    3) 3 x 3 matrix of double:
        [ [1.1, 3.3, 5.5],
          [2.2, 4.4, 6.6],
          [3.3, 5.5, 7.7] ]
        = [a] [uint8:3] [D] [uint8:3] [float64:1.1] [3.3] [5.5] [2.2] [4.4] [6.6] [3.3] [5.5] [7.7]
    
    4) 200 x 300 x 3 RGB Image:
        = [a] [uint8:200] [a] [uint8:300] [u] [uint8:3] ...data...
    
    5) 4 x 3 table with 3 colums, 1 line of text ("lon", "lat", "alt") and 3 lines of doubles:
        [ ["lon", "lat", "alt"],
          [1.1,   3.3,   5.5],
          [2.2,   4.4,   6.6],
          [3.3,   5.5,   7.7] ]
        = [[] [a] [uint8:3] [3] [lon] [lat] [alt] [D] [uint8:3] ...data... []]
    
    6) Repeated values inside an array:
        [12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 10]
        = [[] [u] [uint8:12] [<] [u] [uint8:0] [>] [u] [uint8:20] [u] [uint8:15] [u] [uint8:10] []]
    
    7) 3 x 2 matrix with struct float64 + string5 + int8:
        [[] [<] [[] [<] [[] [d] [k] [b] [>] [u] [uint8:3] [>] [u] [uint8:2] FLOAT64 [string5] INT8 FLOAT64 [string5] INT8 ...data...
    
    8) Header for partially repeated data ("lon" and "lat" should be stored only once):
        [ {"lon": 30, "lat": 6},
          {"lon": 40, "lat": 8},
          {"lon": 50, "lat": 10},
          {"lon": 60, "lat": 12} ]
        = [[] [{] [<] [!] [3] [lon] [d] [!] [3] [lat] [d] [>] [u] [uint8:4] [float64:30] [float64:6] [float64:40] [float64:8] ...data...

**Comment by Miosss** [comment]
@bitagoras
Well thats complex aproach... But I fail to understand one - those subsets you speak of. You mean that in each transmission/appliance, the data is prefixed by the subset index to define what "grammar" they use? Dialects or something? If so, then it would create huge mess I think when trying to connect two parties to understand each other. Not mentioning the coder complexity.

**Comment by MikeFair** [comment]
@bitagoras 
I read through the spec.

You might want to take a look at the documentation for the .MAT File specification from mathworks:
https://www.mathworks.com/help/pdf_doc/matlab/matfile_format.pdf

Or MessagePack:
https://github.com/msgpack/msgpack/blob/master/spec.md

You can usually make a format that compromises between fast random access and compressed representation, or is good at one, or the other; but the needs of fast random access and efficient storage are largely at odds with each other.

For instance, if I had an object named "barry" deep within this structure, it appears I'd have to load/read through every preceding structure before I can find it as the format has no form of index.  Further, there's no length prefix at the start of each structure, so it's not like I could just read a number, skip that many bytes, read the next number, and keep going (for example, the way in which repeating elements are handled requires that I would have to actually read through the structure to figure out how many bytes it contained to find the start of the next data structure after it to see if it was "barry").
---

It's not so much about a straight encoding of the data structure, while that's a challenge it's pretty solvable; the real challenge is having good methods for descriptions of higher order data structures (aka classes).  

For instance, dynamically being able to define a new "type" identifier that represented a class and then reusing that typeId throughout the rest of the file.

For instance, an array of objects that contain a string, an array of 10 elements, and an object with 6 fields (one of which is an object with a variable number of fields) as values.  How many times does redundant field name information and array redefinition need to occur?  Can they have null values?

From my reading of the spec, it appears OBJECTs and ARRAYs aren't available as an ELEMENT.

I also didn't see a way to distinguish a NULL inside an array.
---

If you're working with dense matrices of GBs of data, unless there's some kind of predictable/repeatable aspect to the data, it's just not going to get any smaller and I'd recommend you to use SQLite or Postgres for your use case.

SQLite meets many of your criteria and if treated right, can work well for large databases:
http://stackoverflow.com/questions/784173/what-are-the-performance-characteristics-of-sqlite-with-very-large-database-file

I think PostgreSQL probably meets them all but is no longer just a file based system.

If you would still really like to write your own code, then I'd encourage you to write a python library class that acts like an API into the database, that behind the scenes hides away all the file access.
If you have the control I'd encourage you to partition the data up into multiple files  for faster access.

---
Lastly, is the format for transferring over the network another system or for reading back into the same program.  This makes a difference because tools like diff/rsync/unison can be used to synchronize large files over a network by just copying the text/blocks that have changed.  This is good if you use a format that uses line based or block aligned data structures.

As for UBJ, I'll let others comment on that.
Hope that is useful to you!

**Comment by bitagoras** [comment]
@Miosss: I did not think about how to transmitt the information of the format. Maybe by the extension UBJ like ".JSON.UBJ". A generall viewer should also be able to show which subset a given file has. The names are a cleaner way than a mess of different versions (as usual).

I give you an example: the mat-files of newer matlab-versions are a specification subset of HDF5. This means, every full HDF5 library can read mat files automatically. On the other hand Matlab does not have to support all HDF5 features (what is really complicated to support all HDF5 features).

@MikeFair: I am not interested in just another file format for a specific application. I would like to have one really universal file format to get rid of all the others. It is the same what can be observed in the programming languages. Take Python for example. Python has become very mighty and versatile. It is simple for small programms but offers also many language features like classes for large projects. It is not just another scripting language. It replaces matlab, perl, gnuplot, mathematica and unifies the needs of all scripting languages. There are more than 6000 programming languages in the world, b.t.w. more than human spoken languages (about 5000). I only need two programming languages, one compiler language and one for interactive scripting (even this will be unified one day). I use Python as my universal tool since 10 years for numerics, data plotting, system control and automation. It saved me so much time, to have this ONE universal software.

In formats it is even worse. Formats exist more than programmers. Today a collegue gave me another file format I never saw before with satellite coordinates. You won't believe in how many different ugly formats you can write satellite coordinates to files. A nightmare. I am happy to have sometimes HDF5 that offers some standard in science. But it is not common in other areas. It is too complicated and too restricted to numerical applications. You can't produce a very simple file without having a library. This is where a universal binary notation would be good. For simple data you do not need a library. But in case you can take advantage of more features for applications such as databases, logging, large scientific data. That is what I am looking for. A format for both simple and very large and complex data. A Format were you can decide if your data should be small, fast or random access.

The specification is not finished, there may be bugs in it. I also removed the size and offset information again. It was not perfect and I have to think more about it.

A VALUE can be also an OBJECT or ARRAY. ELEMENT is just a non-container type.

**Comment by thebuzzmedia** [comment]
@bitagoras this is a very impressive piece of work, however it is an entirely separate specification -- I'm going to close this issue because I don't think it helps to have it buried here, but I would strongly encourage you to post this specification publicly somewhere (github?) and begin getting community feedback as you develop it into something more complete.

Great work!

**Comment by bitagoras** [comment]
Supplement to this discussion: After I gave up this idea for a while, I started now (1.5 years later) a new project called [Universal-Binary-Notation](https://github.com/bitagoras/Universal-Binary-Notation) on Github. I found a strategy to keep the grammar very simple compared to what I had last year, almost as simple as UBJSON, but not compatible. These complex features I needed can now be described on a meta level so that the simple grammar is not affected. This means that the parser can stay unchanged when new features are added to the format.
